src/Compilers/Core/Portable/CommandLine/AnalyzerConfig.cs

```diff
@@ -0,0 +1,485 @@
+            /// </summary>
+            public ImmutableArray<Diagnostic> Diagnostics { get; }
+
+            internal AnalyzerConfigOptionsResult(
+                ImmutableArray<TreeOptions> treeOptions,
+                ImmutableArray<AnalyzerOptions> analyzerOptions,
+                ImmutableArray<Diagnostic> diagnostics)
+            {
+                TreeOptions = treeOptions;
+                AnalyzerOptions = analyzerOptions;
+                Diagnostics = diagnostics;
+            }
+        }
+
+        /// <summary>
+        /// Takes a list of paths to source files and a list of AnalyzeConfigs and produces a
+        /// resultant dictionary of diagnostic configurations for each of the source paths.
+        /// Source paths are matched by checking if they are members of the language recognized by
+        /// <see cref="AnalyzerConfig.Section.Name"/>s.
+        /// </summary>
+        /// <param name="sourcePaths">
+        /// Absolute, normalized paths to source files. These paths are expected to be normalized
+        /// using the same mechanism used to normalize the path passed to the path paramater of 
+        /// <see cref="AnalyzerConfig.Parse(string, string)"/>. Source files will only be considered
+        /// applicable for a given <see cref="AnalyzerConfig"/> if the config path is an ordinal
+        /// prefix of the source path.
+        /// </param>
+        /// <param name="analyzerConfigs">
+        /// Parsed AnalyzerConfig files. The <see cref="AnalyzerConfig.NormalizedDirectory"/>
+        /// must be an ordinal prefix of a source file path to be considered applicable.
+        /// </param>
+        /// <returns>
+        /// Diagnostic and analyzer option arrays, where the array indices correspond
+        /// to the options for the source path at the same index.
+        /// </returns>
+        public static AnalyzerConfigOptionsResult GetAnalyzerConfigOptions<TStringList, TACList>(
+            TStringList sourcePaths,
+            TACList analyzerConfigs)
+            where TStringList : IReadOnlyList<string>
+            where TACList : IReadOnlyList<AnalyzerConfig>
+        {
+            // Since paths are compared as ordinal string comparisons, the least nested
+            // file will always be the first in sort order
+            if (!analyzerConfigs.IsSorted(DirectoryLengthComparer))
+            {
+                throw new ArgumentException(
+                    "Analyzer config files must be sorted from shortest to longest path",
+                    nameof(analyzerConfigs));
+            }
+
+            var diagnosticBuilder = ArrayBuilder<Diagnostic>.GetInstance();
+            var allTreeOptions = ArrayBuilder<TreeOptions>.GetInstance(sourcePaths.Count);
+            var allAnalyzerOptions = ArrayBuilder<AnalyzerOptions>.GetInstance(sourcePaths.Count);
+
+            var allRegexes = PooledDictionary<AnalyzerConfig, ImmutableArray<Regex>>.GetInstance();
```

Comment:
Can pass analyzerConfigs.Count as an argument here.

Reply:
PooledDictionary doesn't have a minimum size parameter.

Reply:
Well that's disappointing. 

---
In reply to: [228286932](https://github.com/dotnet/roslyn/pull/30752#discussion_r228286932) [](ancestors = 228286932)